railgun/internal/proxy/seeder/seeder.go

```diff
@@ -0,0 +1,245 @@
+// Seeder
+// -----------------------------------------------------------------------------
+
+// Seeder is an object which can perform pre-fetch seed rounds to cache
+// supported objects into the proxy cache server.
+type Seeder struct {
+	ingressClassName string
+
+	logger logrus.FieldLogger
+	prx    proxy.Proxy
+
+	restCFG  *rest.Config
+	kc       *kubernetes.Clientset
+	kongc    *clientset.Clientset
+	knativec *kubernetes.Clientset // FIXME - TODO
+}
+
+// New provides a need *Seeder object.
+func New(logger logrus.FieldLogger, restCFG *rest.Config, prx proxy.Proxy, ingressClassName string) (*Seeder, error) {
+	kc, err := kubernetes.NewForConfig(restCFG)
+	if err != nil {
+		return nil, err
+	}
+	kongc, err := clientset.NewForConfig(restCFG)
+	if err != nil {
+		return nil, err
+	}
+	return &Seeder{
+		ingressClassName: ingressClassName,
+
+		logger: logger,
+		prx:    prx,
+
+		restCFG: restCFG,
+		kc:      kc,
+		kongc:   kongc,
+		// FIXME - knative
+	}, nil
+}
+
+// Seed lists all supported API types, filters them to make sure the
+// object is supported (e.g. uses the managers' ingress.class) and then
+// pulls fresh copies of the object from the Kubernetes API to (re)seed
+// the proxy cache. This is required to deal with situations where events
+// get lost by controllers due to networking failures, poorly timed controller
+// pod restarts, e.t.c.
+func (s *Seeder) Seed(ctx context.Context) error {
+	// FIXME - namespace filtration
+	objs := make([]client.Object, 0)
+
+	s.logger.Info("fetching supported core kubernetes objects")
+	coreObjs, err := s.fetchCore(ctx)
+	if err != nil {
+		return err
+	}
+	objs = append(objs, coreObjs...)
+
+	s.logger.Info("fetching supported kong kubernetes objects")
+	kongObjs, err := s.fetchKong(ctx)
+	if err != nil {
+		return err
+	}
+	objs = append(objs, kongObjs...)
+
+	// FIXME - knative.Ingress
+
+	if len(objs) < 1 {
+		s.logger.Info("seed round successful: there were no objects that needed to be cached")
+		return nil
+	}
+
+	if err := s.prx.UpdateObjects(objs...); err != nil {
+		return err
+	}
+	s.logger.Infof("seed round successful: %d objects were added to the cache", len(objs))
+
+	return nil
+}
+
+// -----------------------------------------------------------------------------
+// Seeder - Controller Runtime - Runnable Implementation
+// -----------------------------------------------------------------------------
+
+func (s *Seeder) Start(ctx context.Context) error {
+	return s.Seed(ctx)
+}
+
+// -----------------------------------------------------------------------------
+// Seeder - Private Methods
+// -----------------------------------------------------------------------------
+
+// fetchCore fetches a fresh list of all Kubernetes objects and filters them according
+// to the *Seeder's configured s.ingressClassName and returns the result.
+func (s *Seeder) fetchCore(ctx context.Context) ([]client.Object, error) {
```

Comment:
note: I'm considering whether these functions simply need to be generated via our controller generator.

Reply:
I've decided I am going to do this: it will help reduce the burden of adding new APIs.

Reply:
4eb0acf9d1dab0db7dd99796945ddd46578de98e

Reply:
Actually: decided against it after all there are many permutations and the amount of code present isn't enormous.